src/librbd/ManagedLock.cc

```diff
@@ -0,0 +1,588 @@
+    break;
+  }
+}
+
+template <typename I>
+typename ManagedLock<I>::Action ManagedLock<I>::get_active_action() const {
+  assert(m_lock.is_locked());
+  assert(!m_actions_contexts.empty());
+  return m_actions_contexts.front().first;
+}
+
+template <typename I>
+void ManagedLock<I>::complete_active_action(State next_state, int r) {
+  assert(m_lock.is_locked());
+  assert(!m_actions_contexts.empty());
+
+  ActionContexts action_contexts(std::move(m_actions_contexts.front()));
+  m_actions_contexts.pop_front();
+  m_state = next_state;
+
+  m_lock.Unlock();
+  for (auto ctx : action_contexts.second) {
+    ctx->complete(r);
+  }
+  m_lock.Lock();
+
+  if (!is_transition_state() && !m_actions_contexts.empty()) {
+    execute_next_action();
+  }
+}
+
+template <typename I>
+bool ManagedLock<I>::is_shutdown_locked() const {
+  assert(m_lock.is_locked());
+
+  return ((m_state == STATE_SHUTDOWN) ||
+          (!m_actions_contexts.empty() &&
+           m_actions_contexts.back().first == ACTION_SHUT_DOWN));
+}
+
+template <typename I>
+void ManagedLock<I>::send_acquire_lock() {
+  assert(m_lock.is_locked());
+  if (m_state == STATE_LOCKED) {
+    complete_active_action(STATE_LOCKED, 0);
+    return;
+  }
+
+  ldout(m_cct, 10) << dendl;
+  m_state = STATE_ACQUIRING;
+
+  if (!m_watcher->is_registered()) {
+    lderr(m_cct) << "watcher not registered - delaying request" << dendl;
+    m_state = STATE_WAITING_FOR_REGISTER;
+    return;
+  }
+  uint64_t watch_handle = m_watcher->get_watch_handle();
+  m_cookie = ManagedLock<I>::encode_lock_cookie(watch_handle);
+
+  pre_acquire_lock_handler(new FunctionContext([this](int r) {
+    using managed_lock::AcquireRequest;
+    AcquireRequest<I>* req = AcquireRequest<I>::create(m_ioctx, m_watcher,
+        m_work_queue, m_oid, m_cookie,
+        util::create_context_callback<
+          ManagedLock<I>, &ManagedLock<I>::handle_acquire_lock>(this));
+    m_work_queue->queue(new C_SendLockRequest<AcquireRequest<I>>(req), 0);
+  }));
+}
+
+template <typename I>
+void ManagedLock<I>::handle_acquire_lock(int r) {
+  ldout(m_cct, 10) << ": r=" << r << dendl;
+
+  if (r == -EBUSY || r == -EAGAIN) {
+    ldout(m_cct, 5) << ": unable to acquire exclusive lock" << dendl;
+  } else if (r < 0) {
+    lderr(m_cct) << ": failed to acquire exclusive lock:" << cpp_strerror(r)
+               << dendl;
+  } else {
+    ldout(m_cct, 5) << ": successfully acquired exclusive lock" << dendl;
+  }
+
+  State next_state = (r < 0 ? STATE_UNLOCKED : STATE_LOCKED);
+
+  Mutex::Locker locker(m_lock);
+  post_acquire_lock_handler(r, new FunctionContext([this, next_state](int r) {
+    assert(m_lock.is_locked());
+
+    if (r < 0 && next_state == STATE_LOCKED) {
+      // release_lock without calling pre and post callbacks
+      using managed_lock::ReleaseRequest;
+      ReleaseRequest<I>* req = ReleaseRequest<I>::create(m_ioctx, m_watcher,
+          m_work_queue, m_oid, m_cookie,
+          new FunctionContext([this, r](int ret) {
```

Comment:
These lambdas are getting pretty large and complex -- I think, in general, it would read cleaner to just create helper methods that ```create_context_callback``` can just wrap.